[REVIEW] added deps for `gtest` and `gmock` to `conda` environments and `CMakeLists.txt` (#5732)
rapidsai/cudf: cpp/CMakeLists.txt

if(BUILD_TESTS)
    include(CTest)
    find_package(GTest)
    

>     if(NOT GTEST_ROOT)

Reviewer: ```suggestion
    if(NOT GTEST_FOUND)
```

>         include(ConfigureGoogleBenchmark)

Reviewer: Wrong Configure script? Should be for gtest

        set(GTEST_ROOT "${CMAKE_BINARY_DIR}/googletest")

>     endif(GTEST_ROOT)

Reviewer: ```suggestion
    endif(NOT GTEST_FOUND)
```

    add_subdirectory(${CMAKE_SOURCE_DIR}/tests)
    
    set(GTEST_INCLUDE_DIR "${GTEST_ROOT}/install/include")
    set(GTEST_LIBRARY_DIR "${GTEST_ROOT}/install/lib")

>     link_directories(${GTEST_ROOT}/install/lib/)

Reviewer: ```suggestion
    link_directories(${GTEST_LIBRARY_DIR})
```

    include_directories(${GTEST_INCLUDE_DIR})

>     if(NOT GTEST_FOUND)

Reviewer: Does `GTEST_FOUND` get set by the `ConfigureGoogleTest` script?

Author: Yes.

        set(BUILD_TESTING OFF)

>     endif()

Reviewer: ```suggestion
    endif(NOT GTEST_FOUND)
```

>     if(NOT GTEST_FOUND)
>         set(BUILD_TESTING OFF)
>     endif(NOT GTEST_FOUND)
> 
>     message(STATUS "Google C++ Testing Framework (Google Test) was found? ${GTEST_FOUND} - ${GTEST_ROOT}")

Reviewer 2: I don't like the potentially confusing non-grammatical "Gtest was found?" message. ("I can haz cheezburger?")

Instead, use an if / else like we had before.

```suggestion
    if(GTEST_FOUND)
        message(STATUS "Google C++ Testing Framework (Google Test) found in ${GTEST_ROOT}")
    else()
        message(AUTHOR_WARNING "Google C++ Testing Framework (Google Test) not found: automated tests are disabled.")
        set(BUILD_TESTING OFF)
    endif(NOT GTEST_FOUND)
```

if(BUILD_TESTS)
    include(CTest)
    find_package(GTest)
    

>     if(NOT GTEST_ROOT)

Reviewer: ```suggestion
    if(NOT GTEST_FOUND)
```

if(BUILD_TESTS)
    include(CTest)
    find_package(GTest)
    
    if(NOT GTEST_ROOT)
        include(ConfigureGoogleTest)
        set(GTEST_ROOT "${CMAKE_BINARY_DIR}/googletest")
    endif(NOT GTEST_FOUND)

    add_subdirectory(${CMAKE_SOURCE_DIR}/tests)
    
    set(GTEST_INCLUDE_DIR "${GTEST_ROOT}/install/include")
    set(GTEST_LIBRARY_DIR "${GTEST_ROOT}/install/lib")
    link_directories(${GTEST_LIBRARY_DIR})
    include_directories(${GTEST_INCLUDE_DIR})

    if(GTEST_FOUND)
        message(STATUS "Google C++ Testing Framework (Google Test) found in ${GTEST_ROOT}")

>     else():

Reviewer: ```suggestion
    else()
```

        message(AUTHOR_WARNING "Google C++ Testing Framework (Google Test) not found: automated tests are disabled.")
        set(BUILD_TESTING OFF)

>     endif(NOT GTEST_FOUND)

Reviewer: ```suggestion
    endif(GTEST_FOUND)
```

if(BUILD_TESTS)
    include(CTest)
    find_package(GTest)
    
    if(GTEST_FOUND)

>         set(GTEST_INCLUDE_DIR "$ENV{CONDA_PREFIX}/include/gtest")

Reviewer: We can't assume that gtest is found in a conda environment, could have been installed via apt/yum/spack/etc.

if(BUILD_TESTS)
    include(CTest)
    
    if(GTEST_ROOT)
        message(STATUS "Google C++ Testing Framework (Google Test) found in ${GTEST_ROOT}")
        add_subdirectory(${CMAKE_SOURCE_DIR}/tests)
    else()
        message(AUTHOR_WARNING "Google C++ Testing Framework (Google Test) not found, it will be built from source.")
        include(ConfigureGoogleBenchmark)
        add_subdirectory(${CMAKE_SOURCE_DIR}/tests)
        set(GTEST_ROOT "${CMAKE_BINARY_DIR}/googletest")

>     endif(GTEST_ROOT)

Author: Ok. I follow. That should work for both this and benchmarks as long as I'm not setting REQUIRED.

if(BUILD_TESTS)
    include(CTest)
    find_package(GTest)
    
    if(GTEST_ROOT)

>         message(STATUS "Google C++ Testing Framework (Google Test) found in ${GTEST_ROOT}")

Reviewer: Can move this status message to outside of these find / build statements to below them and replace the current status message.

Author: Like just carry a status variable and push the message once?

Reviewer: There is no `STATUS` variable, `STATUS` is just a message type here, and payload of the message is converying `GTEST_ROOT` which is set on either finding the `gtest` package or on building it, so may as well convey the status message after both had run. If you wanted to add a different message per whether we find gtest or build it from source I think that would be great.

Author: A+

        add_subdirectory(${CMAKE_SOURCE_DIR}/tests)
    else()
        message(AUTHOR_WARNING "Google C++ Testing Framework (Google Test) not found, it will be built from source.")
        include(ConfigureGoogleBenchmark)

>         add_subdirectory(${CMAKE_SOURCE_DIR}/tests)

Reviewer: I think we need to run this regardless of whether we find or build gtest

Author: The include? I disagree. I can break it up into conditionals but it does all the checkout/build/install at include. I could split the checkout/build/install into a macro like the Blazing example. I'm not sure why specifically we'd need to do that though. I don't see anything in the module that isn't covered otherwise.

Reviewer: Just the `add_subdirectory` command. That's what adds and configures the unit tests, so we presumably want to run that regardless of whether we find an existing package or build gbench ourselves.

if(BUILD_TESTS)
    include(CTest)

>     find_package(GTest REQUIRED)

Reviewer 2: So what happens if I'm not in a conda env that has gtest. We no longer include ConfigureGoogleTest and so it fails? 

We need to fall back to fetching and building gtest (as before) when it is not found.

Author: OK. I'll check and if fail use the existing configuration.

Reviewer: You can follow the `Arrow` section above where we also have this pattern, and I'd be interested in handling `GTest` similarly in its own section

Reviewer: We should also similarly handle google benchmark in the same way, there's conda packages here: https://anaconda.org/conda-forge/benchmark

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    find_package(benchmark)

    if(NOT benchmark_FOUND)
        include(ConfigureGoogleBenchmark)
        set(benchmark_ROOT "${CMAKE_BINARY_DIR}/googlebenchmark")
    endif(benchmark_FOUND)
    add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)

> 

Reviewer: ```suggestion

    set(GBENCH_INCLUDE_DIR "${benchmark_ROOT}/install/include")
    set(GBENCH_LIBRARY_DIR "${benchmark_ROOT}/install/lib")
    link_directories(${GBENCH_LIBRARY_DIR}
    include_directories(${GBENCH_INCLUDE_DIR})
    
    if(NOT benchmark_FOUND)
        set(BUILD_BENCHMARKS off)
    endif(NOT benchmark_FOUND)
```

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    find_package(benchmark)

    if(NOT benchmark_FOUND)
        include(ConfigureGoogleBenchmark)
        set(benchmark_ROOT "${CMAKE_BINARY_DIR}/googlebenchmark")
    endif(benchmark_FOUND)
    add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)

    set(GBENCH_INCLUDE_DIR "${benchmark_ROOT}/install/include")
    set(GBENCH_LIBRARY_DIR "${benchmark_ROOT}/install/lib")

>     link_directories(${GBENCH_LIBRARY_DIR}

Reviewer: ```suggestion
    link_directories(${GBENCH_LIBRARY_DIR})
```

    include_directories(${GBENCH_INCLUDE_DIR})
    
    if(NOT benchmark_FOUND)
        set(BUILD_BENCHMARKS off)
    endif(NOT benchmark_FOUND)

>     message(STATUS "Google C++ Benchmarking Framework (Google Benchmark) was found? ${benchmark_FOUND} - ${benchmark_ROOT}")

Reviewer 2: See "I can haz cheezburger?" suggestion above and apply here as well. :)

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    find_package(benchmark)

    if(NOT benchmark_FOUND)
        include(ConfigureGoogleBenchmark)
        set(benchmark_ROOT "${CMAKE_BINARY_DIR}/googlebenchmark")

>     endif(benchmark_FOUND)

Reviewer: ```suggestion
    endif(NOT benchmark_FOUND)
```

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    find_package(benchmark)

    if(NOT benchmark_FOUND)
        include(ConfigureGoogleBenchmark)
        set(benchmark_ROOT "${CMAKE_BINARY_DIR}/googlebenchmark")
    endif(benchmark_FOUND)
    add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)

    set(GBENCH_INCLUDE_DIR "${benchmark_ROOT}/install/include")
    set(GBENCH_LIBRARY_DIR "${benchmark_ROOT}/install/lib")
    link_directories(${GBENCH_LIBRARY_DIR}
    include_directories(${GBENCH_INCLUDE_DIR})
    
    if(benchmark_FOUND)
        message(STATUS "Located Google C++ Benchmarking Framework (Google Benchmark) found in ${benchmark_ROOT}")

>     else():

Reviewer: ```suggestion
    else()
```

Author: Fixed in commit pushed at the same time you made this comment.

        message(AUTHOR_WARNING "Located Google C++ Benchmarking Framework (Google Benchmark) not found: automated tests are disabled.")
        set(BUILD_BENCHMARKS off)

>     endif(NOT benchmark_FOUND)

Reviewer: ```suggestion
    endif(benchmark_FOUND)
```

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    find_package(benchmark)

    if(benchmark_FOUND)
        set(GBENCH_INCLUDE_DIR "$ENV{CONDA_PREFIX}/include/benchmark")

>         set(GBENCH_LIBRARY_DIR "$ENV{CONDA_PREFIX}/lib/cmake/benchmark")

Reviewer: Same here

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    find_package(benchmark)

    if(benchmark_FOUND)

>         message(STATUS "Google C++ Benchmarking Framework (Google Benchmark) found in ${benchmark_ROOT}")

Reviewer: Similarly move status message outside of the find / build if statements

Author: After further review, I'm unsure of the reasoning behind this if just storing in a variable. Should I just make one message that conditionally changes based off the `_FOUND` var or something? Is this for readability? The type of message changes with context. I don't quite follow.

Reviewer: I think the idea is the user should get two messages:
1. Whether an existing package is found that we're going to use, or we're building the package from source ourselves
2. Where said package that we found / built is that we'll be using

        add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)
    else()
        message(AUTHOR_WARNING "Google C++ Benchmarking Framework (Google Benchmark) not found, it will be built from source.")
        include(ConfigureGoogleBenchmark)

>         add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)

Reviewer: I think we need to run this regardless of whether we find or build gbench

Author: My response for this is the same as above relating to GTEST.

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    find_package(benchmark)
    
    if(GBENCH_ROOT)
        message(STATUS "Google C++ Benchmarking Framework (Google Benchmark) found in ${GBENCH_ROOT}")

>         add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)

Reviewer: So instead of `GBENCH_ROOT`, it will search in `benchmark_ROOT` in the `find_package` call. If the package is found it will then set a `benchmark_FOUND` variable, along with variables for the library / include dirs (unsure exactly what they will be named).

So instead of checking `GBENCH_ROOT` here, we can do something along the lines of `if(NOT benchmark_FOUND)`  to indicate we need to build it from source using our configuration module.

Author: Right, which is done below so I have checks for it after it installs as well. But if someone does not have a conda (or other) install and I check for GBENCH/GTEST_FOUND instead of GBENCH_ROOT before install it fails out. So in this configuration, it checks for the root, if found, continues. If not found, installs, sets up includes and some vars (definitely required in this pattern) then checks for GTEST_FOUND at the bottom in case the install failed, if failed in the install step for some reason disable benchmarks or tests so there are not conflicts with other steps.

    else()
        message(AUTHOR_WARNING "Google C++ Benchmarking Framework (Google Benchmark) not found, it will be built from source.")
        include(ConfigureGoogleBenchmark)
        add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)
        set(benchmark_ROOT "${CMAKE_BINARY_DIR}/googlebenchmark")
    endif(benchmark_ROOT)

    set(benchmark_INCLUDE_DIR "${benchmark_ROOT}/install/include")
    set(benchmark_LIBRARY_DIR "${benchmark_ROOT}/install/lib" "${benchmark_ROOT}/install/lib64")
    include_directories(${benchmark_INCLUDE_DIR})

    message(STATUS "benchmark_ROOT: " ${benchmark_ROOT})

    if(benchmark_FOUND)
        message(STATUS "Google C++ Benchmarking Framework (Google Benchmark) found in ${benchmark_ROOT}")
    else()
        message(AUTHOR_WARNING "Google C++ Benchmarking Framework (Google Benchmark) not found: automated tests are disabled.")
        set(BUILD_BENCHMARKS OFF)

>     endif()

Reviewer: I think this chunk can be removed now? We either find an installed version or build it from source ourselves.

Author: I think it still serves a purpose. The first checks for the root even if not found. In my tests I've found the if I do `benchmark_FOUND` only or `benchmark_ROOT` and there is no `conda` package or other linked install and it does the install, then the installation can fail to link for the subsequent processes without explicitly setting and doing a check. Also the bottom check if the install failed ensures that `BUILD_BENCHMARKS` is switched to `OFF` if a failure has occurred preventing that variable from being misused in further compilation. 

I get the idea of only using `benchmark_FOUND` and calling it good, but in my tests I've found that not to perform consistently.

# - add google benchmark --------------------------------------------------------------------------

if(BUILD_BENCHMARKS)
    
    if(benchmark_ROOT)
        message(STATUS "Google C++ Benchmarking Framework (Google Benchmark) found in ${benchmark_ROOT}")
        add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)
    else()
        message(AUTHOR_WARNING "Google C++ Benchmarking Framework (Google Benchmark) not found, it will be built from source.")
        include(ConfigureGoogleBenchmark)
        add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)
        set(benchmark_ROOT "${CMAKE_BINARY_DIR}/googlebenchmark")
    endif(benchmark_ROOT)

>     find_package(benchmark)

Reviewer: This should be moved above the `if(benchmark_ROOT)` similar to what you have for `gtest`

if(BUILD_BENCHMARKS)

    if(GBENCH_ROOT)
        message(STATUS "Google C++ Benchmarking Framework (Google Benchmark) found in ${GBENCH_ROOT}")
        add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)
    else()
        message(AUTHOR_WARNING "Google C++ Benchmarking Framework (Google Benchmark) not found, it will be built from source.")
        include(ConfigureGoogleBenchmark)
        add_subdirectory(${CMAKE_SOURCE_DIR}/benchmarks)
        set(GBENCH_ROOT "${CMAKE_BINARY_DIR}/googlebenchmark")

>     endif(GBENCH_ROOT)

Reviewer: Should be able to do the same here with `find_package(benchmark)`

Author: This step is required if it's not found to build it. As I removed that include from the top of the bench setup and put it here. I had issues with `findpackage` working after checkout and building even when linked. With both gtest and benchmarks. `findpackage` would claim it couldn't find it even with hints. Are you saying the whole thing could be replaced or just the variable setting and subdirectory?

Reviewer: Did you install the `benchmark` conda package? Otherwise your existing environment files don't have google benchmark being installed.

Author: Added. But I was also solving for situations when benchmark has not been installed by conda.
